### DISCUSSION TITLE
Translate Cloudlog

### DISCUSSION THREAD

**Discussion by magicbug** [magicbug/Cloudlog#795] [General]
2021 is the year for translation Cloudlog, this, of course, means all the text within Cloudlog needs to be broken up so that it can be done, I've created a simple Kaban using Github Projects option at https://github.com/magicbug/Cloudlog/projects/7

At the moment Dashboard is done, but every other area in Cloudlog needs similar treatment it will be easier in some areas as certain words are used all over the place so they just need to be defined once.

I'm personally going to handle QSO and Contest entry areas to make sure the code is handled correctly on these parts but if you want to grab another area then PRs "per" screen will be accepted, I'm going to suggest that its per screen or section to make verifying it easier.

Once the code changes have been done for English we will start looking for other language translations to just bare with us while this happens

**Comment by magicbug** [comment]
**Contesting Page Ready for Translation**

Just to say that contesting interface has now had the translation system implemented for English, it uses words from the contesting lang file and the general words file.

**Comment by magicbug** [comment]
**QSO Entry Ready for Translation**

Next up is the QSO Entry screen within Cloudlog, think I've caught all the parts so should be able to translate to any language now, took 40mins to do this screen alone!

**Comment by magicbug** [comment]
If you wanting to get involved adding languages other than English, then language files are stored in /application/language you'll notice English has a pile of files these all need translating, you'll notice German already has one file with some support.

I'm using the language system within Codeigniter see https://codeigniter.com/userguide3/libraries/language.html for details.

**Comment by 112bzo** [comment]
I'am now making the translation to Dutch,

**Comment by francescocozzi** [comment]
Hi @magicbug you made a great job! I'm running Cloudlog on a Digital Ocean Ubuntu droplet like a charme. I'm at a good point with the Italian translation. In next days I would push you files with a commit. Ciao. Frank IZ7KHR.

**Reply by magicbug** [reply]
Thanks Frank much appreciated

**Comment by LZ2ISI** [comment]
I can help with a Bulgarian translation. Ivan LZ2ISI.

**Reply by magicbug** [reply]
Yes please 🙂

**Comment by 112bzo** [comment]
I have translated the following file into Dutch : lotw_lang

`<?php

defined('BASEPATH') OR exit('No direct script access allowed');

$lang['lotw_short'] = 'LoTW';
$lang['lotw_title'] = 'Logbook of the World';
$lang['lotw_title_available_cert'] = 'Beschikbare certificaten';
$lang['lotw_title_information'] = 'Informatie';
$lang['lotw_title_upload_p12_cert'] = 'Upload Logbook of the World .p12 Certificate';
$lang['lotw_title_export_p12_file_instruction'] = 'Export .p12 File Instructions';
$lang['lotw_title_adif_import'] = 'ADIF Importeren';
$lang['lotw_title_adif_import_options'] = 'Importeer opties';

$lang['lotw_beta_warning'] = 'Houd er rekening mee dat LoTW Sync BETA is, zie wiki voor hulp.';
$lang['lotw_no_certs_uploaded'] = 'U moet enkele LoTW p12-certificaten uploaden om dit gebied te gebruiken.';

$lang['lotw_date_created'] = 'Aanmaak datum';
$lang['lotw_date_expires'] = 'Verloop datum';
$lang['lotw_status'] = 'Status';
$lang['lotw_options'] = 'Opties';
$lang['lotw_valid'] = 'Geldig';
$lang['lotw_expired'] = 'Verlopen';
$lang['lotw_not_synced'] = 'Niet gesychroniseerd';

$lang['lotw_certificate_dxcc'] = 'Certificate DXCC';
$lang['lotw_certificate_dxcc_help_text'] = 'Certificate DXCC. voorbeeld: Schotland';

$lang['lotw_input_a_file'] = 'Upload een bestand';

$lang['lotw_upload_exported_adif_file_from_lotw'] = 'Upload het geëxporteerde ADIF-bestand van LoTW vanuit de <a href="https://p1k.arrl.org/lotwuser/qsos?qsoscmd=adif" target="_blank">Download Report</a> Area, to mark QSOs as confirmed on LOTW.';
$lang['lotw_upload_type_must_be_adi'] = 'Logbestanden moeten het bestandstype hebben .adi';

$lang['lotw_pull_lotw_data_for_me'] = 'Laad LoTW-gegevens voor mij';
$lang['lotw_import_missing_qsos_text'] = 'Importeer ontbrekende QSOs in het logboek. Call en gridsquare worden gecontroleerd om te proberen het juiste profiel te vinden om de QSO in te importeren. Indien niet gevonden, wordt de QSO overgeslagen.';

$lang['lotw_report_download_overview_helptext'] ='Cloudlog gebruikt de LoTW-gebruikersnaam en het wachtwoord die zijn opgeslagen in uw gebruikersprofiel om een rapport van LoTW voor u te downloaden. Het rapport Cloudlog-downloads hebben alle bevestigingen sinds de gekozen datum, of sinds je laatste LoTW-bevestiging (opgehaald uit je log), tot nu toe.';

// Buttons
$lang['lotw_btn_lotw_import'] = 'LoTW Import';
$lang['lotw_btn_upload_certificate'] = 'Upload Certificate';
$lang['lotw_btn_delete'] = 'Verwijder';
$lang['lotw_btn_manual_sync'] = 'handmatig synchroniseren';
$lang['lotw_btn_upload_file'] = 'Upload bestand';
$lang['lotw_btn_import_matches'] = 'Importeer LoTW hits';

// P12 Export Text
$lang['lotw_p12_export_step_one'] = 'Open TQSL &amp; ga naar de Callsign Certificates Tab';
$lang['lotw_p12_export_step_two'] = 'Klik met de rechtermuisknop op het gewenste Roepnaam';
$lang['lotw_p12_export_step_three'] = 'Klik op "Save Callsign Certificate File" en voeg geen wachtwoord toe';
$lang['lotw_p12_export_step_four'] = 'Upload bestand hier onder.';

$lang['lotw_confirmed'] = 'Dit QSO wordt bevestigd op LoTW';
`

**Comment by kitobelix** [comment]
Hi! Translated to spanish too.

I've found that there are some texts that doesn't appear in those files. I will dig deeper to see if I can find them.
[cloudlog-spanish.zip](https://github.com/magicbug/Cloudlog/files/8286291/cloudlog-spanish.zip)

Bye!

**Reply by phl0** [reply]
I reviewed this zip file and got some questions:
- Which branch are the changes based on?
- Is there a reason why the English parts are just commented out? I vote for simply removing it

**Reply by kitobelix** [reply]
Hi! Thanks for the review. About point 2, yes. I vote to remove it too.

I've never set this out to be part of a pull request because of the incompleteness of this set of files and my ignorance about the rest of the translation files located at ./application/language/*

Sorry if this is inconvenient but I think that my file should only be taken as reference to begin translating CL into spanish. 
I'm working at building a usable and easy to deploy docker image with this translations included. 

Again thank you for the review! I really appreciate your time.

**Comment by kitobelix** [comment]
Nevermind, I've found the rest at /application/language hehe

**Reply by R1BLH** [reply]
Hi!
I can help with a Russian translation. My local instance is partially translated already.
R1BLH

**Comment by LarsThunberg** [comment]
Hi, 
I would like to add Swedish translation. What is the correct way to work, this one?

- Make a new branch from Master
- Add the translation folder / files
- Make a Pull Request

//Lars SM0TGU

**Reply by phl0** [reply]
> Hi, I would like to add Swedish translation. What is the correct way to work, this one?

Basically that is the correct way. But due to V2 being published (maybe today) I would recommend to branch from station_logbooks rather than master. 
I have made a little script which outputs missing language tags compared to English. I could provide you with that if needed.

**Reply by LarsThunberg** [reply]
> Basically that is the correct way. But due to V2 being published (maybe today) I would recommend to branch from station_logbooks rather than master.

I will wait after the release of V2, then I can take a branch from Master. I will take the English files as base so no tags will be missing.

**Reply by LarsThunberg** [reply]
But it seams I can not create a new branch... I can do the translation local and maybe someone can help to make a PR?

**Reply by phl0** [reply]
You need to first fork the whole repository into your account. Then you can checkout station_logbooks branch and checkout a branch from there. No need to wait for publication of V2. I can support with github stuff if needed.

**Reply by R1BLH** [reply]
>  I have made a little script which outputs missing language tags compared to English. I could provide you with that if needed.

Could you share this script, please?

**Reply by phl0** [reply]
For sure. Here you go:

```
#!/bin/bash
lang="german"
echo
echo "Missing translation tags:"
for file in application/language/english/*.php; do
  for i in $(grep -h "\$lang\['.*\]" $file | sed -e "s/\$lang\['\(.*\)'\] = .*/\1/"); do
    grep -rq "$i" application/language/$lang/*;
    if [ $? != "0" ]; then
       echo "$file: $i missing!";
    fi;
  done
done
exit 0;
```

**Reply by R1BLH** [reply]
> For sure. Here you go:
Thank you!

**Comment by R1BLH** [comment]
Hi! Translated to russian all lines except hardcoded.
Also i made static buttons and menu items to callsign lookup on the qrz.ru.

**Comment by R1BLH** [comment]
@magicbug could you please make  language tags for menus (and other strings to be translated)? 

![2022-11-10_16-24-02](https://user-images.githubusercontent.com/16667173/201103382-fdfe70c9-8aed-4245-b793-3794e04f239b.png)

**Reply by magicbug** [reply]
Yep will get that done

**Reply by R1BLH** [reply]
@magicbug 
> Yep will get that done

I checked untraslated tags in the last version of the Cloudlog.

_Missing translation tags:
application/language/english/general_words_lang.php: gen_hamradio_pota_reference missing!
application/language/english/general_words_lang.php: gen_hamradio_pota missing!
application/language/english/lotw_lang.php: lotw_expiring missing!
application/language/english/lotw_lang.php: lotw_cert_expiring missing!
application/language/english/lotw_lang.php: lotw_cert_expired missing!_

Could you add more tags to translate please? (menus and other items)?

**Comment by R1BLH** [comment]
@magicbug please advise how to add new language file (for example to translate menus and OQRS)? I tried add new language file with translation and replace hardcoded text lines to something like `<?php $this->lang->line('constant_name'); ?>`, but this gave no result. It looks like i need to declare new language file somewhere.

**Reply by AndreasK79** [reply]
I have it working locally. I modified the controller in the system to always load the menu translation.
When I'm done, and tested, it will be merged into dev. When that is done, you can add translation.

**Reply by R1BLH** [reply]
Thank you :) I'm not skilled enough in web programming to understand code structure on the fly.

**Reply by AndreasK79** [reply]
I have added a pull request for the menu. You can take the menu translation from that and translate if you want.

Regarding translations:

1. Make a file for the translation (example oqrs_lang.php)
2. Replace the strings in the view with `<?php echo $this->lang->line('constant_name'); ?>`
3. Load the language file in the controller constructor `$this->lang->load(array('languagefile'));`

The menu was a special case, since it is used everywhere.

**Reply by phl0** [reply]
I think it should be

```
<?php echo $this->lang->line('constant_name'); ?>
```

i.e. an `echo` missing :)

**Reply by AndreasK79** [reply]
@phl0 good thing I have a proofreader :) Edited now.

**Reply by phl0** [reply]
Just wanted to prevent @cats-shadow being puzzled why it would not work :)

**Reply by R1BLH** [reply]
> Just wanted to prevent @cats-shadow being puzzled why it would not work :)

Thanks! :)
The web-programming is more tricky thing for me when hardware design/construction. :) But i'm learning. :)

**Reply by R1BLH** [reply]
> I have added a pull request for the menu. You can take the menu translation from that and translate if you want.

Thanks! //Still have no idea how to get PR from another fork to my fork. Git is really a puzzle.

>     3. Load the language file in the controller constructor `$this->lang->load(array('languagefile'));`

This moment i missed. Found it. Thanks!

**Reply by AndreasK79** [reply]
The pull request is merged. Just pull latest dev, and you will have the necessary file to translate.

**Reply by R1BLH** [reply]
> The pull request is merged. Just pull latest dev, and you will have the necessary file to translate.

Done!

**Reply by R1BLH** [reply]
> Just wanted to prevent @cats-shadow being puzzled why it would not work :)

I'm confused again:
The plural form in English is simple: just add 's' to the end of the word (see the line of code below):
` span.innerText = span.textContent = '('+(grid_four_count-grid_four_confirmed_count)+' grid square'+(grid_four_count-grid_four_confirmed_count != 1 ? 's' : '')+') ';
`
The situation in Russian is more variable. Let's take an example from above: the word "grid square" is "квадрат" in Russian.
if the number ends with 2 or 3 or 4: add "а" -- "квадрата";
if the number ends with 5 or 6, or 7, or 8, or 9, or 0: "ов" should be added -- "квадратов".

So code above must be modified to be dependable from selected language. How to do it?

**Reply by R1BLH** [reply]
Another problem. The statistics are generated by the script /assets/js/sections/statistics.js. It uses hardcoded text constants that are displayed on the screen.
What's the correct way to do the translation in this case (as far as I understand, js runs in the client's browser, php runs on the server)?

**Reply by AndreasK79** [reply]
@cats-shadow I see you never got an answer. The solution would be to add the translations needed to variables in the controller, and output them as javascript variables in the view. Then use the variables in the javascript.

**Comment by HB9HIL** [comment]
Dear Hams,

In the past few days, I have put a lot of effort into adding basic multilanguage support to various parts of Cloudlog. I have included language lines for every available language. Not everything is finished yet, but I am continuously working on it. 😄 

Now, all the enthusiastic hams who wish to help with translation and possess the necessary language skills can fork Cloudlog, translate the snippets in `application/language/*`, and create a Pull Request.

If you intend to add new language snippets to the language files, please keep in mind these two points:
- Is your branch up to date with the dev branch of magicbug's Cloudlog? If not, you may run into a branch conflict.
- Add the new language snippet to every available language in the same language file to ensure that all languages are in sync. The command `sed` will help you with that.

Thank you for your assistance! :-)

**Reply by magicbug** [reply]
@HB9HIL First of all thank you for going through doing the translations it's a massive job 

The more languages Cloudlog can support the better!

**Reply by HB9HIL** [reply]
Adding Support for other languages is no big problem. Finding People who translate that is more an issue 😆 I'm not sure if Google Translate is good enough for that.

**Reply by magicbug** [reply]
That is always the problem, if theres any specific langs people want Cloudlog to support that it doesn't already please of course let us know we might be able todo something about it if a volunteer cannot be found

**Comment by DanielBausset** [comment]
Hello,

I'll be able to provide french translations, if nobody is already on it :)

73
